Hack Week Project: Added support for in app updates

anitaa1990 opened the pull request:
Fixes #1272 by adding support for updating the app without having to go to Google Play. There are two types of in app updates: **FLEXIBLE** and **IMMEDIATE**. 

### Changes:
- Added the Google Play Core library to the app.
- Added a flag to `gradle.properties` to check if app update type is flexible or immediate.  
- Created a new activity class called `AppUpdateActivity` which handles all the app update logic. This activity is currently inherited by `MainActivity`. The logic is to check in app updates happen only **after** the user is logged in to the app. (This was mostly because I didn’t think it was good user experience to display an upgrade dialog as soon as the app is opened 🤷‍♀)
- If a user already cancels the update, it would be annoying to display the update dialog again and again. So added logic to NOT display this update dialog if the user has cancelled it for the current version. So if a new version is provided in Google Play, then the dialog will be displayed again. In the future, we could add an option somewhere in the Settings or menu to update the app (if the user wishes to do so, after cancelling the update dialog).

### Notes:
In order to support both types of updates, the following property has been added to the `gradle.properties`. This is just for testing purposes to display both types of upgrades. We would need to modify the logic in the future, to display both types of updates, in various screens, depending on the scenario involved or the app update version. 
```
# The type of in app update FLEXIBLE = 0; IMMEDIATE = 1
wc.in_app_update_type = 0
```

### Screenshots:
(LEFT: Flexible Update, RIGHT: Immediate Update)
<img width="300" src="https://user-images.githubusercontent.com/22608780/61705375-47d4b280-ad63-11e9-8424-aa0cbddf7a22.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/61705407-54f1a180-ad63-11e9-81b4-330796cf819d.gif">

#### Flexible in app update dialog (with and without Wifi):
<img width="300" src="https://user-images.githubusercontent.com/22608780/61705557-997d3d00-ad63-11e9-9147-53bf9418c287.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/61705555-984c1000-ad63-11e9-90f2-2afd6037819d.png">

#### Immediate in app update dialog (with and without Wifi):
<img width="300" src="https://user-images.githubusercontent.com/22608780/61705618-bca7ec80-ad63-11e9-814a-b8d71b5df017.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/61705619-bdd91980-ad63-11e9-90e5-0e51020971f5.png">

### Testing:
While the feature itself is straightforward to implement, the testing process is quite long and irksome. I apologise in advance for the number of scenarios listed below for testing 😬 . There are two options for end to end testing: 
 - Create a release build signed with the same key used to sign the app before uploading to the Play Store. We would need to upload the app to the the internal testing track.
 - Another option is to decrease the `versionCode` and `versionName` in the `build.gradle` and create a release build. This would prompt an update in the app (since the `versionCode` is less than the `versionCode` in Google Play).

I went with Option number 2 by following the below steps to test each of the scenarios listed below:
- [x] Decrease the version code to 55 and create a release build.
- [x] Install the release build
- [x] Verify that the upgrade dialog will be displayed only after login.

#### FLEXIBLE upgrade - Cancel update scenario:
- [x] Click on `No Thanks` button.
- [x] Notice that you are able to use the app. 
- [x] Close the app and open it again. Verify that the upgrade dialog is  not displayed again.

#### FLEXIBLE upgrade - Upgrade using Wifi:
- [x] Uninstall the app and install a fresh one from the changes in this PR.
- [x] Click on `UPDATE` button.
- [x] Notice that you are able to use the app. 
- [x] Verify that once the upgrade is successful, a snackBar is displayed. (This is not dismissible and will be displayed till the Install button is clicked).
- [x] Click on Install in the SnackBar and verify that the app is installed successfully. Once install is over, verify that the Google Play app version and the app version in the device is the same.

#### FLEXIBLE upgrade - Upgrade using Mobile Network only:
- [x] Uninstall the app and install a fresh one from the changes in this PR.
- [x] Verify that only mobile network is on and Wifi is switched off. 
- [x] Click on `Now (over any network)` button (for mobile network)
- [x] Notice that you are able to use the app. 
- [x] Verify that once the upgrade is successful, a snackBar is displayed. (This is not dismissible and will be displayed till the Install button is clicked).
- [x] Click on Install in the SnackBar and verify that the app is installed successfully. Once install is over, verify that the Google Play app version and the app version in the device is the same.

##### Variant test for the above scenario:
- [x] Repeat Steps 1, 2 & 3 from above. 
- [x] **Close the app** after clicking on the Upgrade now button. 
- [x] Open the app again.
- [x]  Verify that once the upgrade is successful, a snackBar is displayed. (This is not dismissible and will be displayed till the Install button is clicked).
- [x] Click on Install in the SnackBar and verify that the app is installed successfully. Once install is over, verify that the Google Play app version and the app version in the device is the same.

#### FLEXIBLE upgrade - Upgrade only when Wifi is available:
- [x] Verify that only mobile network is on and Wifi is switched off. 
- [x] Select “When Wifi is available” radio button and click on `UPDATE` button.
- [x] Notice that you are able to use the app. 
- [x] Enable Wifi from the device notification bar. 
- [x] Verify that the upgrade is started once Wifi is connected. 
- [x] Verify that once the upgrade is successful, a snackBar is displayed. 
- [x] Click on Install in the SnackBar and verify that the app is installed successfully. Once install is over, verify that the Google Play app version and the app version in the device is the same.

##### Variant test for the above scenario:
- [x] Verify that only mobile network is on and Wifi is switched off. 
- [x] Click on “Wait for Wifi” button.
- [x] Notice that you are able to use the app. Close the app. 
- [x] Enable Wifi from the device settings. 
- [x] Open the app again and notice that you are able to use the app.
- [x] Verify that once the upgrade is successful, a snackBar is displayed.  
- [x] Click on Install in the SnackBar and verify that the app is installed successfully. Once install is over, verify that the Google Play app version and the app version in the device is the same. 

### IMMEDIATE upgrade 
- [x] Change the value of `wc.in_app_update_type` in grade.properties from 0 to 1.
- [x] Decrease the version code to 55 and create a build.
- [x] Install the build
- [x] Verify that the upgrade dialog will be displayed only after login.

#### IMMEDIATE  - Cancel update scenario:
- [x]  Click on cross icon at the top right of the page.
- [x] Notice that you are able to use the app. Close the app and open it again. Verify that the upgrade dialog is  not displayed again.

IMMEDIATE upgrade - Upgrade using Wifi OR Mobile Network:
- [x] Uninstall the app and install a fresh one from the changes in this PR.
- [x] Click on `UPDATE` button.
- [x] Notice that you are NOT able to use the app. 
- [x] Verify that once the upgrade is successful, the app is installed and restarted automatically.
- [x] Once install is over, verify that the Google Play app version and the app version in the device is the same.

##### Variant test for the above scenario:
- [x] Repeat Steps 1 & 2 from above. 
- [x] **Close the app** after clicking on the Upgrade now button. 
- [x] Open the app again.
- [x] Verify that the update progress dialog is displayed again and you are not able to use the app. 
- [x] Verify that once the upgrade is successful, the app is installed and restarted automatically. Verify that the Google Play app version and the app version in the device is the same.

#### IMMEDIATE upgrade - Upgrade using Mobile Network only:
- [x] Verify that only mobile network is on and Wifi is switched off. 
- [x] Select “When Wifi is available” radio button and click on UPDATE.
- [x] Notice that you are NOT able to use the app. 
- [x] Enable Wifi from the **device notification bar**. 
- [x] Verify that the upgrade is started and the progress dialog is displayed. 
- [x] Verify that once the upgrade is successful, the app is installed and restarted automatically.
- [x] Once install is over, verify that the Google Play app version and the app version in the device is the same.

##### Variant test for the above scenario:
- [x] Verify that only mobile network is on and Wifi is switched off. 
- [x] Click on “Wait for Wifi” button.
- [x] Notice that you are NOT able to use the app.
- [x] **Close the app** and enable Wifi from the device settings. 
- [x] Open the app again. Notice that the upgrade is started and the progress dialog is displayed. 
- [x] Verify that once the upgrade is successful, the app is installed and restarted automatically.
- [x] Once install is over, verify that the Google Play app version and the app version in the device is the same.
 
#### No Internet available when checking for app updates
 - [x] The `appUpdateInfo.isUpdateTypeAllowed` will always return `false`. So the update dialog will not be displayed.

#### Internet available when checking for app updates but not available AFTER update dialog displayed 
 - [x] If the download has started before the internet is switched off: The download is actually in a `Pending` state and it’s waiting for the Wifi to be turned back on so it can resume the download. I did some testing on whether we get a callback during this pending event and although there is an option called `InstallState.PENDING`, we don’t seem to get a callback for this particular scenario.   
     - [x] In the case of `IMMEDIATE` update: the UI displays a waiting for Wifi screen and also a notification is displayed in the notification bar stating the same.
     - [x] In the case of `FLEXIBLE` update: the user is still allowed to use the app but the notification is displayed in the notification bar to let them know that the download is still pending. It will resume automatically when the Wifi becomes available. 
 - [x] If the download has NOT started before the internet is switched off: The update is in a `FAILED` state so we display the retry snackbar here.



Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

cc @AmandaRiu and @nbradbury, could I trouble you both for a review on this please? Thank you!!!

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt:
```diff
@@ -43,7 +43,9 @@ object AppPrefs {
-        NUM_TIMES_MARK_ALL_NOTIFS_READ_SNACK_SHOWN
+        NUM_TIMES_MARK_ALL_NOTIFS_READ_SNACK_SHOWN,
+        // The app update for this version was cancelled by the user
+        CANCELLED_APP_VERSION_CODE,
```
WDYT about also tracking when the user did _not_ cancel the update? I think it would be helpful to compare the two events.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/UIMessageResolver.kt:
```diff
@@ -22,6 +22,25 @@ interface UIMessageResolver {
     val snackbarRoot: ViewGroup
 
+    /**
+     * Create and return a snackbar displaying the provided message and a RESTART button.
```
This comment could be more clear about the purpose of the snackbar. Can we change it to state that it's related to updating the app to a new version?

anitaa1990 replied:
👍 Done in 06b32ca

nbradbury commented:
> * So added logic to NOT display this update dialog if the user has cancelled it for the current version. So if a new version is provided in Google Play, then the dialog will be displayed again.

💯

nbradbury commented on WooCommerce/src/main/res/values/strings.xml:
```diff
@@ -35,6 +35,9 @@
     <string name="undo">Undo</string>
+    <string name="install">Install</string>
+    <string name="update_downloaded">Woo has downloaded an update</string>
+    <string name="update_failed">App update failed. Tap to retry</string>
```
np: Since there's a period after "App update failed" there should also be one after "Tap to retry."

AmandaRiu replied:
For additional context to Nick's comment, see the [skip unnecessary punctuation](https://material.io/design/communication/writing.html#principles) section in the Material Design Guidelines. The basics are:
- If there is a single sentence, don't use punctuation (unless it's a question)
- If multiple, use punctuation.

anitaa1990 replied:
Thank you @AmandaRiu for the reference! I've updated the string text for failure scenario in e501aea

AmandaRiu replied:
@anitaa1990 I noticed you removed the "tap to retry" text, does that mean we no longer support it? If we do, then I'd change it back to what the text previously was and just add a period at the end.

AmandaRiu replied:
NM - I just saw one of the screenshots and the button says everything needed 👍

nbradbury commented:
@anitaa1990 You weren't kidding about the testing process being "long and irksome!" 😄 

Everything looks & works great with one exception. I let the app show the update dialog and started the update, and while it was updating I killed the network connection. I expected to see the snackbar stating that the update failed, but it never appeared. I know that's an edge case and would be fine to merge regardless, but maybe this is something that's simple to handle?

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/AppUpgradeActivity.kt:
```diff
@@ -0,0 +1,219 @@
+                        // if this version > that the stored app version code.
+                        val availableAppVersionCode = appUpdateInfo.availableVersionCode()
+                        val storedAppVersionCode = AppPrefs.getCancelledAppVersionCode()
+                        if (availableAppVersionCode > storedAppVersionCode) {
```
I wonder if we should have some sort of timed expiration for the canceled update. A user may be in the middle of something and decide not to update at that very moment, but may want to update later. Is it possible to add a "Do not ask again" checkbox (default = false) when we ask the user if they would like to apply an update? If they don't check that box, then we will continue to ask every time they open the app.

nbradbury replied:
> Is it possible to add a "Do not ask again" checkbox (default = false) when we ask the user if they would like to apply an update?

That update dialog is Google's and I don't think it's customizable. I know we don't want that dialog popping up all the time, but we could limit it to once a day or once every other day.

FWIW, though, I'm fine with keeping it the way it is. We do such frequent updates that I don't think it's a big deal if we don't alert the user until the next version.

anitaa1990 replied:
Yup, as @nbradbury mentioned, since the update dialog is Google's we won't be able to edit it. 

As I mentioned in the description though, I thought of adding an option somewhere in the Settings page or perhaps in the menu underneath the `Help` option to update the app if the user cancelled the current update and would like the option to do so later. (I'm not sure if those are the current places for this option. cc @kyleaparker  on your thought here). 

But I thought of tackling this separately seeing as we would need to incorporate the design changes. WDYT?

AmandaRiu replied:
OIC, ok. Yeah a setting option sounds like a great idea 👍 We can do it in a separate PR though so it doesn't hold up this one.

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/AppUpgradeActivity.kt:
```diff
@@ -0,0 +1,219 @@
+                        val availableAppVersionCode = appUpdateInfo.availableVersionCode()
+                        val storedAppVersionCode = AppPrefs.getCancelledAppVersionCode()
+                        if (availableAppVersionCode > storedAppVersionCode) {
+                            if (isAppUpdateImmediate()) {
```
IIRC, and immediate update is a _required update_. If that's correct, shouldn't this block exist outside the check for cancelation?

anitaa1990 replied:
Yup, that's what I thought too. But there is an option to cancel the update in both types of dialog. So by `IMMEDIATE` update, maybe Google meant user will not be able to use the app while the update is going on. But they can cancel it at any time. 
<img width="300" src="https://user-images.githubusercontent.com/22608780/61705618-bca7ec80-ad63-11e9-814a-b8d71b5df017.png">

Good news is, we do get a callback if the user cancels so we would be able to handle this scenario by blocking the user from using the app. For now, I just store the fact that the user cancelled the dialog in `SharedPreferences` and let them use the app.

AmandaRiu replied:
Ok, thank you for the explanation. Google explained it a bit differently during Google IO 👍

anitaa1990 commented:
Thank you for the quick review @nbradbury and @AmandaRiu 😄Sorry for the extensive testing Nick!

> I let the app show the update dialog and started the update, and while it was updating I went into airplane mode. I expected to see the snackbar stating that the update failed, but it never appeared.

Nice catch! In this case, there are a couple of scenarios that can take place.
- If there is no Internet available when checking for app updates: The `appUpdateInfo.isUpdateTypeAllowed` will always return false. So the update dialog will not be displayed.
- If there is Internet available when checking for app updates but not available **AFTER update dialog displayed and UPGRADE button is clicked** -> (the case you mentioned)
  - **If the download has started before the internet is switched off:** The download is actually in a Pending state and it’s waiting for the Wifi to be turned back on so it can resume the download. In the case of IMMEDIATE update: the UI displays a waiting for Wifi screen and also a notification is displayed in the notification bar stating the same. In the case of FLEXIBLE update: As you mentioned, the user is still allowed to use the app but the notification is displayed in the notification bar to let them know that the download is still pending. It will resume automatically when the Wifi becomes available. I did some testing for both the scenarios to check if we get a callback during this pending event and although there is an option called `InstallState.PENDING`, we don’t seem to get a callback for this particular scenario. 
(LEFT: IMMEDIATE update .   RIGHT: FLEXIBLE update)
<img width="300" src="https://user-images.githubusercontent.com/22608780/61767515-2676d380-ae02-11e9-80bb-ba9ee58e213b.png"> .  <img width="300" src="https://user-images.githubusercontent.com/22608780/61767784-fd0a7780-ae02-11e9-9bb9-6ee650ffccba.png">

  - **If the download has NOT started before the internet is switched off:** The update is in a FAILED state so in this scenario, we display the retry snackbar here.
<img width="300" src="https://user-images.githubusercontent.com/22608780/61767893-478bf400-ae03-11e9-82d5-0f4e46bd92e4.png"> .  

I've updated this scenario to the PR description as part of one of the testing scenarios. 

@nbradbury @AmandaRiu, I've addressed the nitpicks and PR is ready for another round 👍
